File: Software/RF2DFieldSolver/Scenarios/Scenario.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;

namespace RF2DFieldSolver.Scenarios
{
    public abstract class Scenario : Form
    {
        protected List<Parameter> parameters;
        protected string name;
        protected TableLayoutPanel ui;

        public Scenario()
        {
            parameters = new List<Parameter>();
            ui = new TableLayoutPanel();
            ui.Dock = DockStyle.Fill;
            Controls.Add(ui);

            var buttonBox = new FlowLayoutPanel();
            buttonBox.Dock = DockStyle.Bottom;
            Controls.Add(buttonBox);

            var acceptButton = new Button { Text = "OK" };
            acceptButton.Click += (sender, e) => OnAccept();
            buttonBox.Controls.Add(acceptButton);

            var cancelButton = new Button { Text = "Cancel" };
            cancelButton.Click += (sender, e) => OnCancel();
            buttonBox.Controls.Add(cancelButton);

            var autoAreaCheckBox = new CheckBox { Text = "Auto Area" };
            ui.Controls.Add(autoAreaCheckBox);

            var xLeftLabel = new Label { Text = "X Left" };
            ui.Controls.Add(xLeftLabel);
            var xLeftTextBox = new TextBox();
            ui.Controls.Add(xLeftTextBox);

            var xRightLabel = new Label { Text = "X Right" };
            ui.Controls.Add(xRightLabel);
            var xRightTextBox = new TextBox();
            ui.Controls.Add(xRightTextBox);

            var yTopLabel = new Label { Text = "Y Top" };
            ui.Controls.Add(yTopLabel);
            var yTopTextBox = new TextBox();
            ui.Controls.Add(yTopTextBox);

            var yBottomLabel = new Label { Text = "Y Bottom" };
            ui.Controls.Add(yBottomLabel);
            var yBottomTextBox = new TextBox();
            ui.Controls.Add(yBottomTextBox);
        }

        protected abstract ElementList CreateScenario();
        protected abstract Image GetImage();

        private void OnAccept()
        {
            foreach (var parameter in parameters)
            {
                var control = ui.Controls[parameter.Name] as TextBox;
                if (control != null)
                {
                    parameter.Value = double.Parse(control.Text);
                }
            }

            var list = CreateScenario();
            var topLeft = new PointF(float.Parse(ui.Controls["X Left"].Text), float.Parse(ui.Controls["Y Top"].Text));
            var bottomRight = new PointF(float.Parse(ui.Controls["X Right"].Text), float.Parse(ui.Controls["Y Bottom"].Text));
            ScenarioCreated?.Invoke(topLeft, bottomRight, list);
            Close();
        }

        private void OnCancel()
        {
            Close();
        }

        public static List<Scenario> CreateAll()
        {
            var scenarios = new List<Scenario>
            {
                new Microstrip(),
                new CoplanarMicrostrip(),
                new DifferentialMicrostrip(),
                new CoplanarDifferentialMicrostrip(),
                new Stripline(),
                new CoplanarStripline(),
                new DifferentialStripline(),
                new CoplanarDifferentialStripline()
            };

            foreach (var scenario in scenarios)
            {
                scenario.SetupParameters();
            }

            return scenarios;
        }

        public void SetupParameters()
        {
            foreach (var parameter in parameters)
            {
                var label = new Label { Text = parameter.Name + ":" };
                var textBox = new TextBox { Name = parameter.Name, Text = parameter.Value.ToString() };
                ui.Controls.Add(label);
                ui.Controls.Add(textBox);
            }

            Text = name + " Setup Dialog";

            var pictureBox = new PictureBox
            {
                Image = GetImage(),
                Dock = DockStyle.Fill,
                SizeMode = PictureBoxSizeMode.StretchImage
            };
            ui.Controls.Add(pictureBox);
        }

        public string Name => name;

        public event Action<PointF, PointF, ElementList> ScenarioCreated;

        public class Parameter
        {
            public string Name { get; set; }
            public string Unit { get; set; }
            public string Prefixes { get; set; }
            public int Precision { get; set; }
            public double Value { get; set; }
        }
    }
}

File: Software/RF2DFieldSolver/Scenarios/scenario.cpp
#include "scenario.h"
#include "ui_scenario.h"

#include "CustomWidgets/siunitedit.h"

#include "microstrip.h"
#include "stripline.h"
#include "coplanarmicrostrip.h"
#include "coplanarstripline.h"
#include "differentialmicrostrip.h"
#include "coplanardifferentialmicrostrip.h"
#include "differentialstripline.h"
#include "coplanardifferentialstripline.h"

Scenario::Scenario(System::Windows::Forms::Form^ parent) :
    Form(parent),
    ui(gcnew Ui::Scenario)
{
    ui->setupUi(this);

    ui->parameters->setLayout(gcnew System::Windows::Forms::TableLayoutPanel);
    ui->buttonBox->CancelButton = gcnew System::Windows::Forms::Button();
    ui->buttonBox->CancelButton->Click += gcnew System::EventHandler(this, &Scenario::reject);
    ui->buttonBox->AcceptButton = gcnew System::Windows::Forms::Button();
    ui->buttonBox->AcceptButton->Click += gcnew System::EventHandler(this, [=](System::Object^ sender, System::EventArgs^ e){
        // update parameters
        for(unsigned int i=0;i<parameters.size();i++) {
            auto layout = static_cast<System::Windows::Forms::TableLayoutPanel^>(ui->parameters->Layout);
            auto entry = static_cast<SIUnitEdit^>(layout->GetControlFromPosition(1, i));
            *parameters[i].value = entry->value();
        }

        auto list = createScenario();
        scenarioCreated(System::Drawing::PointF(ui->xleft->value(), ui->ytop->value()), System::Drawing::PointF(ui->xright->value(), ui->ybottom->value()), list);
        this->DialogResult = System::Windows::Forms::DialogResult::OK;
        this->Close();
    });
    ui->autoArea->Checked = true;
    ui->xleft->setUnit("m");
    ui->xleft->setPrefixes("um ");
    ui->xleft->setPrecision(4);
    ui->xleft->setValue(-3e-3);

    ui->xright->setUnit("m");
    ui->xright->setPrefixes("um ");
    ui->xright->setPrecision(4);
    ui->xright->setValue(3e-3);

    ui->ytop->setUnit("m");
    ui->ytop->setPrefixes("um ");
    ui->ytop->setPrecision(4);
    ui->ytop->setValue(3e-3);

    ui->ybottom->setUnit("m");
    ui->ybottom->setPrefixes("um ");
    ui->ybottom->setPrecision(4);
    ui->ybottom->setValue(-1e-3);
}

Scenario::~Scenario()
{
    delete ui;
}

QList<Scenario *> Scenario::createAll()
{
    QList<Scenario*> ret;
    ret.push_back(gcnew Microstrip());
    ret.push_back(gcnew CoplanarMicrostrip());
    ret.push_back(gcnew DifferentialMicrostrip());
    ret.push_back(gcnew CoplanarDifferentialMicrostrip());
    ret.push_back(gcnew Stripline());
    ret.push_back(gcnew CoplanarStripline());
    ret.push_back(gcnew DifferentialStripline());
    ret.push_back(gcnew CoplanarDifferentialStripline());

    for each (auto s in ret) {
        s->setupParameters();
    }
    return ret;
}

void Scenario::setupParameters()
{
    auto layout = static_cast<System::Windows::Forms::TableLayoutPanel^>(ui->parameters->Layout);
    for each (auto p in parameters) {
        auto label = gcnew System::Windows::Forms::Label();
        label->Text = p.name + ":";
        auto entry = gcnew SIUnitEdit(p.unit, p.prefixes, p.precision);
        entry->setValue(*p.value);
        layout->Controls->Add(label);
        layout->Controls->Add(entry);
    }
    this->Text = name + " Setup Dialog";

    // show the image
    ui->image->Image = getImage();
}
